File: 08-2-3-Decorator-Solution/Decorators/StopWatchDecorator.cs
namespace Decorator.Decorators
{
    using System;
    using System.Diagnostics;

    public sealed class StopWatchDecorator : ICalculator
    {
        private readonly ICalculator _decoratedCalculator;

        public StopWatchDecorator(ICalculator decoratedCalculator)
        {
            _decoratedCalculator = decoratedCalculator;
        }

        public void Calculate()
        {
            var stopWatch = new Stopwatch();
            stopWatch.Start();

            _decoratedCalculator.Calculate();

            stopWatch.Stop();
            Console.WriteLine($"{nameof(StopWatchDecorator)}: The decorated operation took {stopWatch.Elapsed}");
        }
    }
}

File: 08-DesignPatterns-Java/Solution/Decorator/src/Decorators/StopWatchDecorator.java
package Decorators;

import Calculator.ICalculator;

import java.time.Duration;
import java.time.Instant;

public final class StopWatchDecorator implements ICalculator {
    private final ICalculator _decoratedCalculator;

    public StopWatchDecorator(ICalculator decoratedCalculator) {
        _decoratedCalculator = decoratedCalculator;
    }

    @Override
    public void calculate() {
        var starts = Instant.now();
        _decoratedCalculator.calculate();
        var ends = Instant.now();
        System.out.println(getClass().getName() + ": The decorated operation took " + Duration.between(starts, ends));
    }
}
